include/tvm/script/printer/traced_object_functor.h

```diff
@@ -0,0 +1,170 @@
+  using IsDispatchFunction =
+      typename std::is_convertible<TCallable, std::function<R(TracedObject<TObjectRef>, Args...)>>;
+
+ public:
+  /*!
+   * \brief Call the dispatch function.
+   * \param token The dispatch token.
+   * \param traced_object The traced object.
+   * \param args Other args.
+   *
+   * \return The return value of the dispatch function
+   *
+   * If the TObjectRef isn't registered with the token, it will try to find
+   * dispatch function for TObjectRef with kDefaultDispatchToken.
+   */
+  template <class TObjectRef>
+  R operator()(const String& token, TracedObject<TObjectRef> traced_object, Args... args) const {
+    const runtime::PackedFunc& dispatch_function = dispatch_table::GetDispatchFunction(
+        dispatch_table_, token, traced_object.Get()->type_index());
+    return dispatch_function(traced_object.Get(), traced_object.GetPath(),
+                             std::forward<Args>(args)...);
+  }
+
+  /*!
+   * \brief Set the dispatch function
+   * \param token The dispatch token.
+   * \param type_index The TVM object type index for this dispatch function.
+   * \param f The dispatch function.
+   *
+   * This takes a type-erased packed function as input. It should be used
+   * through FFI boundary, for example, registering dispatch function from Python.
+   */
+  TSelf& set_dispatch(String token, uint32_t type_index, runtime::PackedFunc f) {
+    dispatch_table::SetDispatchFunction(&dispatch_table_, token, type_index, std::move(f));
+    return *this;
+  }
+
+  /*!
+   * \brief Set the dispatch function
+   * \param token The dispatch token.
+   * \param f The dispatch function.
+   *
+   * The diaptch function should have signature `R(TracedObject<TObjectRef>, Args...)`.
+   */
+  template <typename TCallable,
+            typename TObjectRef = typename detail::first_arg_type<TCallable>::ObjectRefType,
+            typename = std::enable_if_t<IsDispatchFunction<TObjectRef, TCallable>::value>>
+  TSelf& set_dispatch(String token, TCallable f) {
+    return set_dispatch(token,                                          //
+                        TObjectRef::ContainerType::RuntimeTypeIndex(),  //
+                        runtime::TypedPackedFunc<R(TObjectRef, ObjectPath, Args...)>(
+                            [f](TObjectRef object, ObjectPath path, Args... args) -> R {
+                              return f(MakeTraced(object, path), std::forward<Args>(args)...);
+                            }));
+  }
+  /*!
+   * \brief Set the default dispatch function
+   * \param f The dispatch function.
+   *
+   * Default dispatch function will be used if there is no function registered
+   * with the requested dispatch token.
+   *
+   * Default dispatch function has an empty string as dispatch token.
+   */
+  template <typename TCallable>
+  TSelf& set_dispatch(TCallable f) {
+    return set_dispatch(dispatch_table::kDefaultDispatchToken, std::forward<TCallable>(f));
```

Comment:
(nit) Same comment about `std::forward` here -- we should either have `TCallable&& f` or remove the `std::forward`

Reply:
Updated to use `TCallable&&`